### PR TITLE
Fix pool op doc and class declarations

### PR DESCRIPTION
**Pull request by chengduoZH**
1. Fix pool_op document.
2. In order to reuse PoolOp and PoolMaker in pool_cudnn, I separate the declarations and implementation of the PoolOp and PoolMaker class.

### CODE REVIEW

File: paddle/operators/math/CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,15 @@
 if(WITH_GPU)
-    nv_library(math_function SRCS math_function.cc math_function.cu im2col.cc im2col.cu pooling.cc pooling.cu DEPS cblas device_context operator)
+    nv_library(math_function SRCS math_function.cc math_function.cu im2col.cc im2col.cu DEPS cblas device_context operator)
     nv_test(math_function_test SRCS math_function_test.cc DEPS math_function tensor)
     nv_library(softmax SRCS softmax.cc softmax.cu DEPS operator)
     nv_library(cross_entropy SRCS cross_entropy.cc cross_entropy.cu DEPS operator)
+    nv_library(pooling SRCS pooling.cc pooling.cu DEPS operator)
```
Comment (qingqing01) [inline paddle/operators/math/CMakeLists.txt]: remove the dependence `operator`

File: paddle/operators/math/CMakeLists.txt [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 else()
-    cc_library(math_function SRCS math_function.cc im2col.cc pooling.cc DEPS cblas device_context operator)
+    cc_library(math_function SRCS math_function.cc im2col.cc DEPS cblas device_context operator)
     cc_test(math_function_test SRCS math_function_test.cc DEPS math_function tensor)
     cc_library(softmax SRCS softmax.cc DEPS operator)
     cc_library(cross_entropy SRCS cross_entropy.cc DEPS operator)
+    cc_library(pooling SRCS pooling.cc DEPS operator)
```
Comment (qingqing01) [inline paddle/operators/math/CMakeLists.txt]: same as above

File: paddle/operators/pool_op.cc [hunk 3]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
+
+  auto in_x_dims = ctx->GetInputDim("X");
+
+  std::string pooling_type = ctx->Attrs().Get<std::string>("poolingType");
+  std::vector<int> ksize = ctx->Attrs().Get<std::vector<int>>("ksize");
+  std::vector<int> strides = ctx->Attrs().Get<std::vector<int>>("strides");
+  std::vector<int> paddings = ctx->Attrs().Get<std::vector<int>>("paddings");
+
+  PADDLE_ENFORCE(in_x_dims.size() == 4 || in_x_dims.size() == 5,
+                 "Pooling intput should be 4-D or 5-D");
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: Pooling intput should be 4-D or 5-D

->

The input should be 4-D or 5-D tensor.

File: paddle/operators/pool_op.cc [hunk 4]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
+            "width of feature.");
+
+  AddAttr<std::string>("poolingType",
+                       "PoolingType of pooling operator."
+                       "Str constant equal to 'max' or 'avg'.")
+      .InEnum({"max", "avg"});
+
+  AddAttr<std::vector<int>>(
+      "ksize",
+      "The pooling size(height, width) of pooling operator."
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: The pooling window size

File: paddle/operators/pool_op.cc [hunk 5]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
+                 "Input(X@GRAD) should not be null.");
+  ctx->SetOutputDim(framework::GradVarName("X"), ctx->GetInputDim("X"));
+}
+
+Pool2dOpMaker::Pool2dOpMaker(framework::OpProto *proto,
+                             framework::OpAttrChecker *op_checker)
+    : OpProtoAndCheckerMaker(proto, op_checker) {
+  AddInput(
+      "X",
+      "The input tensor of pooling operator. "
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: specify input type as tensor.

"The input tensor of pooling operator. "

->

"(Tensor) The input tensor of pooling operator. "

File: paddle/operators/pool_op.cc [hunk 6]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
+      "The format of input tensor is NCHW. Where N is batch size, C is the "
+      "number of channels, H and W is the height and width of feature.");
+  AddOutput("Out",
+            "The output tensor of pooling operator."
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: same as above

File: paddle/operators/pool_op.cc [hunk 7]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
+                      // TypedAttrChecker don't support vector type.)
+  AddAttr<bool>(
+      "globalPooling",
+      "Whether to use the globalPooling."
+      "Bool constant equal to false or true."
+      "Default false."
+      "If globalPooling = true, ksize is ignored and need not be specified.")
+      .SetDefault(false);
+  AddAttr<std::vector<int>>("strides",
+                            "Strides(height, width) of pooling operator."
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: The strides(height, width) of pooling window.

File: paddle/operators/pool_op.cc [hunk 8]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
+                            "Default {1,1}.")
+      .SetDefault({1, 1});  // TODO(Chengduo): Add checker. (Currently,
+                            // TypedAttrChecker don't support vector type.)
+  AddAttr<std::vector<int>>("paddings",
+                            "Paddings(height, width) of pooling operator."
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: The zero padding(height, width) size on both sides.

File: paddle/operators/pool_op.cc [hunk 9]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
+      .SetDefault({0, 0});  // TODO(Chengduo): Add checker. (Currently,
+                            // TypedAttrChecker don't support vector type.)
+
+  AddComment(R"DOC(
 The pooling2d operation calculates the output based on
 the input, poolingType and ksize, strides, paddings parameters.
+Input(X) and output(Out) are in NCHW format. Where N is batch size, C is the
+number of channels, H and W is the height and width of feature.
+Parameters(ksize, strides, paddings) are two elements.
+These two elements represent height and width, respectively.
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: Better to give how to calculate the output height/width according to the input height/with, padding, stride size.

The doc of Pytorch is good.

http://pytorch.org/docs/master/nn.html#maxpool2d

File: paddle/operators/pool_op.cc [hunk 10]
```diff
@@ -22,157 +22,158 @@ int OutputSizePool(int input_size, int filter_size, int padding, int stride) {
...
-        .SetDefault({1, 1, 1});  // TODO(Add checker)
-    AddAttr<std::vector<int>>(
-        "paddings",
-        "Paddings(depth, height, width) of pooling operator."
-        "Default {0,0,0}.")
-        .SetDefault({0, 0, 0});  // TODO(Add checker)
-    AddComment(R"DOC(
+}
+
+Pool3dOpMaker::Pool3dOpMaker(framework::OpProto *proto,
```
Comment (qingqing01) [inline paddle/operators/pool_op.cc]: The doc has similar problem with Pool2d.

File: paddle/operators/pool_op.cc (hunk 3 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 5 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 6 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 4 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 7 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 8 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 10 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

File: paddle/operators/pool_op.cc (hunk 9 above)
Comment (chengduoZH) [reply paddle/operators/pool_op.cc]: Done

**Pull request merged**